File: app/src/main/java/com/pr0gramm/app/services/UploadService.java
package com.pr0gramm.app.services;

import android.annotation.SuppressLint;
import android.graphics.Bitmap;
import android.support.annotation.NonNull;

import com.google.common.base.Joiner;
import com.google.common.base.Optional;
import com.google.common.collect.FluentIterable;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableSet;
import com.pr0gramm.app.api.pr0gramm.Api;
import com.pr0gramm.app.feed.ContentType;
import com.pr0gramm.app.services.config.ConfigService;
import com.pr0gramm.app.util.BackgroundScheduler;
import com.squareup.picasso.Picasso;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.Collections;
import java.util.List;
import java.util.Set;

import javax.inject.Inject;
import javax.inject.Singleton;

import okhttp3.MediaType;
import okhttp3.MultipartBody;
import okhttp3.RequestBody;
import okio.BufferedSink;
import retrofit2.adapter.rxjava.HttpException;
import rx.Observable;
import rx.subjects.BehaviorSubject;
import rx.subjects.Subject;

/**
 */
@Singleton
public class UploadService {
    private static final Logger logger = LoggerFactory.getLogger("UploadService");

    private final Api api;
    private final UserService userService;
    private final Picasso picasso;
    private final ConfigService configService;

    @Inject
    public UploadService(Api api, UserService userService, Picasso picasso, ConfigService configService) {
        this.api = api;
        this.userService = userService;
        this.picasso = picasso;
        this.configService = configService;
    }

    private Observable<Long> maxSize() {
        return userService.loginState()
                .take(1)
                .map(state -> state.premium()
                        ? configService.config().maxUploadSizePremium()
                        : configService.config().maxUploadSizeNormal());
    }

    public Observable<Boolean> sizeOkay(File file) {
        return maxSize().map(maxSize -> file.length() < maxSize);
    }

    @SuppressLint("NewApi")
    public Observable<File> downsize(File file) {
        return maxSize().flatMap(maxSize -> Observable.fromCallable(() -> {
            logger.info("Try to scale {}kb image down to max of {}kb",
                    file.length() / 1024, maxSize / 1024);

            Bitmap bitmap = picasso.load(file)
                    .config(Bitmap.Config.ARGB_8888)
                    .centerInside()
                    .resize(2048, 2048)
                    .onlyScaleDown()
                    .get();

            logger.info("Image loaded at {}x{}px", bitmap.getWidth(), bitmap.getHeight());

            // scale down to temp file
            File target = File.createTempFile("upload", "jpg", file.getParentFile());
            target.deleteOnExit();

            Bitmap.CompressFormat format = Bitmap.CompressFormat.JPEG;
            int quality = 90;
            do {
                try (OutputStream output = new FileOutputStream(target)) {
                    logger.info("Compressing to {} at quality={}", format, quality);
                    if (!bitmap.compress(format, quality, output))
                        throw new IOException("Could not compress image data");

                    logger.info("Size is now {}kb", target.length() / 1024);
                }

                // decrease quality to shrink even further
                quality -= 10;
            } while (target.length() >= maxSize && quality > 30);

            logger.info("Finished downsizing with an image size of {}kb", target.length() / 1024);
            return target;
        }));
    }

    private Observable<UploadInfo> upload(File file) {
        if (!file.exists() || !file.isFile() || !file.canRead())
            return Observable.error(new FileNotFoundException("Can not read file to upload"));

        Subject<UploadInfo, UploadInfo> result = BehaviorSubject.create(new UploadInfo(0.f)).toSerialized();

        RequestBody output = new RequestBody() {
            @Override
            public MediaType contentType() {
                try {
                    String type = MimeTypeHelper.guess(file).or("image/jpeg");
                    return MediaType.parse(type);
                } catch (IOException ignored) {
                    return MediaType.parse("image/jpeg");
                }
            }

            @Override
            public long contentLength() throws IOException {
                return file.length();
            }

            @SuppressLint("NewApi")
            @Override
            public void writeTo(BufferedSink sink) throws IOException {
                float length = (float) file.length();
                byte[] buffer = new byte[16 * 1024];

                long lastTime = 0L;
                try (InputStream input = new FileInputStream(file)) {
                    // send first progress report.
                    result.onNext(new UploadInfo(0.f));

                    int len, sent = 0;
                    while ((len = input.read(buffer)) >= 0) {
                        sink.write(buffer, 0, len);
                        sent += len;

                        long now = System.currentTimeMillis();
                        if (now - lastTime >= 100) {
                            lastTime = now;

                            // send progress now.
                            float progress = sent / length;
                            result.onNext(new UploadInfo(progress));
                        }
                    }

                    // tell that the file is sent
                    result.onNext(new UploadInfo(1.f));
                }
            }
        };

        MultipartBody body = new MultipartBody.Builder()
                .setType(MediaType.parse("multipart/form-data"))
                .addFormDataPart("image", file.getName(), output)
                .build();

        long size = file.length();

        // perform the upload!
        api.upload(body)
                .doOnEach(n -> Track.upload(size))
                .map(response -> new UploadInfo(response.getKey(), Collections.emptyList()))
                .subscribeOn(BackgroundScheduler.instance())
                .subscribe(result::onNext, result::onError, result::onCompleted);

        return result.ignoreElements().mergeWith(result);
    }

    private Observable<Api.Posted> post(String key, ContentType contentType, Set<String> tags, boolean checkSimilar) {
        String sfwType = contentType.name().toLowerCase();
        String tagStr = FluentIterable.from(tags)
                .filter(tag -> !INVALID_TAGS.contains(tag.toLowerCase()))
                .append(sfwType)
                .transform(String::trim)
                .filter(tag -> !"sfw".equals(tag.toLowerCase()))
                .join(Joiner.on(","));

        return api.post(null, sfwType, tagStr, checkSimilar ? 1 : 0, key);
    }

    public Observable<UploadInfo> upload(File file, ContentType sfw, Set<String> tags) {
        return upload(file).flatMap(status -> {
            if (status.key != null) {
                return post(status, sfw, tags, true);
            } else {
                return Observable.just(status);
            }
        });
    }

    public Observable<UploadInfo> post(UploadInfo status, ContentType sfw, Set<String> tags, boolean checkSimilar) {
        return post(status.key, sfw, tags, checkSimilar).flatMap(response -> {
            if (response.getSimilar().size() > 0) {
                return Observable.just(new UploadInfo(status.key, response.getSimilar()));

            } else if (response.getError() != null) {
                return Observable.error(new UploadFailedException(response.getError(), response.getReport()));

            } else {
                return Observable.just(new UploadInfo(response.getItemId()));
            }
        });
    }

    /**
     * Checks if the current user is rate limited. Returns true, if the user
     * is not allowed to upload an image right now. Returns false, if the user is
     * allowed to upload an image.
     */
    public Observable<Boolean> checkIsRateLimited() {
        return api.ratelimited().map(v -> false).onErrorResumeNext(error -> {
            if (error instanceof HttpException) {
                // a http error code 403 tells us that the user is rate limited.
                if (((HttpException) error).code() == 403)
                    return Observable.just(true);
            }

            // just forward the error
            return Observable.error(error);
        });
    }

    public static class UploadInfo {
        final String key;
        private final List<HasThumbnail> similar;
        private final long id;
        private final float progress;

        private UploadInfo(long id) {
            this.id = id;
            this.key = null;
            this.progress = -1;
            this.similar = Collections.emptyList();
        }

        private UploadInfo(String key, List<? extends HasThumbnail> similar) {
            this.id = 0;
            this.key = key;
            this.progress = -1;
            this.similar = ImmutableList.copyOf(similar);
        }

        private UploadInfo(float progress) {
            this.id = 0;
            this.key = null;
            this.progress = progress;
            this.similar = Collections.emptyList();
        }

        public boolean isFinished() {
            return id > 0;
        }

        public float getProgress() {
            return progress;
        }

        public long getId() {
            return id;
        }

        public boolean hasSimilar() {
            return similar.size() > 0;
        }

        public List<HasThumbnail> similar() {
            return similar;
        }
    }

    private static final ImmutableSet<Object> INVALID_TAGS = ImmutableSet.of(
            "sfw", "nsfw", "nsfl", "gif", "webm", "sound");

    public static final class UploadFailedException extends Exception {
        @NonNull
        public final Optional<Api.Posted.VideoReport> report;

        public UploadFailedException(String message, @NonNull Optional<Api.Posted.VideoReport> report) {
            super(message);

            this.report = report;
        }

        public String getErrorCode() {
            return getMessage();
        }
    }
}

File: app/src/main/java/com/pr0gramm/app/services/UploadService.kt
package com.pr0gramm.app.services

import android.app.Application
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.net.Uri
import com.pr0gramm.app.Logger
import com.pr0gramm.app.api.pr0gramm.Api
import com.pr0gramm.app.feed.ContentType
import com.pr0gramm.app.seconds
import com.pr0gramm.app.services.config.ConfigService
import com.pr0gramm.app.ui.upload.UploadMediaType
import com.pr0gramm.app.ui.upload.UploadViewModel
import com.pr0gramm.app.ui.views.viewer.MediaUri
import com.pr0gramm.app.util.*
import com.squareup.picasso.Picasso
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.runInterruptible
import okhttp3.MediaType
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.MultipartBody
import okhttp3.RequestBody
import okio.BufferedSink
import retrofit2.HttpException
import java.io.*
import java.util.*
import kotlin.time.ExperimentalTime


/**
 */
class UploadService(
    private val api: Api,
    private val userService: UserService,
    private val picasso: Picasso,
    private val configService: ConfigService,
    private val voteService: VoteService,
    private val applicationContext: Application
) {

    private val logger = Logger("UploadService")

    /**
     * Maximum upload size for the currently signed in user.
     */
    private fun maxUploadSize(type: MediaUri.MediaType): Long {
        val config = configService.config()

        return when (type) {
            MediaUri.MediaType.VIDEO -> config.maxUploadSizeVideo
            else -> config.maxUploadSizeImage
        }
    }

    private val maxUploadPixels: Long
        get() {
            val config = configService.config()
            return if (userService.loginState.premium) config.maxUploadPixelsPremium else config.maxUploadPixelsNormal
        }

    suspend fun sizeOkay(file: File, mediaType: UploadMediaType): Boolean {
        return runInterruptible(Dispatchers.IO) {
            val fileSizeOk = file.length() < maxUploadSize(mediaType.mediaUriType)
            if (!fileSizeOk) {
                return@runInterruptible false
            }

            if (mediaType === UploadMediaType.VIDEO) {
                return@runInterruptible true
            }

            try {
                // get image size, ignore errors
                val opts = BitmapFactory.Options()
                opts.inJustDecodeBounds = true
                BitmapFactory.decodeFile(file.path, opts)

                // check that image size is okay
                opts.outWidth * opts.outHeight <= maxUploadPixels

            } catch (err: Exception) {
                true
            }
        }
    }

    suspend fun shrinkImage(file: File): File {
        return runInterruptible(Dispatchers.IO) {
            val maxUploadSize = maxUploadSize(MediaUri.MediaType.IMAGE)

            logger.info {
                "Try to scale ${file.length() / 1024}kb image down to max of ${maxUploadSize / 1024}kb"
            }

            val bitmap = picasso.load(file)
                .config(Bitmap.Config.ARGB_8888)
                .centerInside()
                .resize(4096, 4096)
                .onlyScaleDown()
                .get()

            logger.info { "Image loaded at ${bitmap.width}x${bitmap.height}px" }

            // scale down to temp file
            val target = File.createTempFile("upload", "jpg", file.parentFile)
            target.deleteOnExit()

            val format = Bitmap.CompressFormat.JPEG
            var quality = 90
            do {
                FileOutputStream(target).use { output ->
                    logger.info { "Compressing to $format at quality=$quality" }
                    if (!bitmap.compress(format, quality, output))
                        throw IOException("Could not compress image data")

                    logger.info { "Size is now ${target.length() / 1024}kb" }
                }

                // decrease quality to shrink even further
                quality -= 10
            } while (target.length() >= maxUploadSize && quality > 30)

            logger.info { "Finished downsizing with an image size of ${target.length() / 1024}kb" }

            // simulate slow operation
            debugOnly { Thread.sleep(1_000) }

            target
        }
    }

    private fun upload(file: File): Flow<State> {
        val states = channelFlow {
            if (!file.exists() || !file.isFile || !file.canRead()) {
                throw FileNotFoundException("Can not read file to upload")
            }

            // prepare the upload
            this.trySend(State.Uploading(progress = 0f)).isSuccess

            val image = RequestBodyWithProgress(file) { state -> this.trySend(state).isSuccess }

            val body = MultipartBody.Builder()
                .setType("multipart/form-data".toMediaTypeOrNull()!!)
                .addFormDataPart("image", file.name, image)
                .build()

            // upload the actual file
            val response = api.upload(body)
            send(State.Uploaded(response.key))

            // track the uploaded file
            Track.upload(size = file.length())
        }

        return states.buffer(16).flowOn(Dispatchers.IO)
    }

    fun post(
        key: String, contentType: ContentType, userTags: Set<String>,
        checkSimilar: Boolean
    ): Flow<State> {

        val contentTypeTag = contentType.name.lowercase(Locale.ROOT)
        val tags = userTags.map { it.trim() }.filter { isValidTag(it) }

        // we can only post 4 extra tags with an upload, so lets add the rest later
        val firstTags: List<String>
        val extraTags: List<String>

        if (contentType == ContentType.SFW) {
            firstTags = tags.take(5)
            extraTags = tags.drop(5)
        } else {
            // we need to add the nsfw/nsfl tag to the list of tags.
            // It looks like the sfw status is getting ignored
            firstTags = listOf(contentTypeTag) + tags.take(4)
            extraTags = tags.drop(4)
        }

        val uploadState = flow {
            val posted = api.post(null, contentTypeTag, firstTags.joinToString(","), checkSimilar.toInt(), key, 1)

            val state = postedToState(key, posted)

            if (state is State.Pending) {
                emitAll(waitOnQueue(state.queue))
            } else {
                emit(state)
            }
        }

        return uploadState.onEach { state ->
            if (state is State.Success && extraTags.isNotEmpty()) {
                ignoreAllExceptions {
                    voteService.createTags(state.id, extraTags)
                }
            }
        }
    }

    private fun postedToState(key: String, posted: Api.Posted): State {
        if (posted.similar.isNotEmpty()) {
            return State.SimilarItems(key, posted.similar)
        }

        val error = posted.error
        if (error != null) {
            return State.Error(error, posted.report)
        }

        if (posted.itemId > 0) {
            return State.Success(posted.itemId)
        }

        posted.queueId?.let { queue ->
            return State.Pending(queue, -1)
        }

        throw IllegalStateException("can not map result to state")
    }

    @OptIn(ExperimentalTime::class)
    private fun waitOnQueue(queue: Long): Flow<State> {
        return flow {
            while (true) {
                kotlinx.coroutines.delay(2.seconds.inMillis)

                // get the current state from the queue - on error just try again.
                val info = try {
                    api.queue(queue)
                } catch (err: Exception) {
                    logger.warn(err) { "Error polling queue" }
                    continue
                }

                val itemId = info.item?.id ?: 0
                when {
                    info.status == "done" && itemId > 0 ->
                        return@flow emit(State.Success(itemId))

                    info.status == "processing" ->
                        emit(State.Processing)

                    else ->
                        emit(State.Pending(queue, info.position))
                }
            }
        }
    }

    sealed class State {
        class Error(val error: String, val report: Api.Posted.VideoReport?) : State()
        class Uploading(val progress: Float) : State()
        class Uploaded(val key: String) : State()
        class SimilarItems(val key: String, val items: List<Api.Posted.SimilarItem>) : State()
        class Success(val id: Long) : State()
        class Pending(val queue: Long, val position: Long) : State()
        object Processing : State()
    }

    fun upload(file: File, sfw: ContentType, tags: Set<String>): Flow<State> {
        return upload(file).transform { state ->
            emit(state)

            if (state is State.Uploaded) {
                emitAll(post(state.key, sfw, tags, true))
            }
        }
    }

    /**
     * Checks if the current user is rate limited. Returns true, if the user
     * is not allowed to upload an image right now. Returns false, if the user is
     * allowed to upload an image.
     */
    suspend fun checkIsRateLimited(): Boolean {
        try {
            api.ratelimited()
            return false
        } catch (err: Throwable) {
            if (err is HttpException && err.code() == 403) {
                return true
            }

            throw err
        }
    }

    suspend fun copyToStaging(source: Uri): Pair<File, UploadMediaType> {
        fun makeTempUploadFile(ext: String): File {
            return File(applicationContext.cacheDir, "upload.${System.currentTimeMillis()}.$ext").also { file ->
                // cleanup if the application exits
                file.deleteOnExit()
            }
        }

        return runInterruptible(Dispatchers.IO) {
            applicationContext.contentResolver.openInputStream(source).use { input ->
                requireNotNull(input) { "Could not open input stream" }

                // read the first few bytes as "header"
                val bytes = BoundedInputStream(input, 512).readBytes()

                // and guess the type, also fail if we couldn't get
                // an extension (and thereby a type)
                val mimeType = MimeTypeHelper.guess(bytes)
                val ext = mimeType
                    ?.let { MimeTypeHelper.extension(it) }
                    ?: throw UploadViewModel.MediaNotSupportedException()

                logger.info { "Creating temporary file with extension '$ext'" }
                val target = makeTempUploadFile(ext)

                FileOutputStream(target).use { output ->
                    output.write(bytes)

                    val maxSize = configService.config().maxUploadSizeVideo + 1
                    val copied = BoundedInputStream(input, maxSize).copyTo(output)
                    logger.info { "Copied ${copied / 1024}kb (max is ${maxSize / 1024}kb" }

                    if (copied == maxSize) {
                        throw IOException("File too large.")
                    }
                }

                val uploadMediaType = when {
                    mimeType.startsWith("video/") -> UploadMediaType.VIDEO
                    else -> UploadMediaType.IMAGE
                }

                Pair(target, uploadMediaType)
            }
        }
    }

    class UploadFailedException(message: String, val report: Api.Posted.VideoReport?) : Exception(message) {
        val errorCode: String get() = message!!
    }

    private class RequestBodyWithProgress(private val file: File, private val publishState: (State) -> Unit) :
        RequestBody() {
        override fun contentType(): MediaType {
            val guessed = runCatching { MimeTypeHelper.guess(file)?.toMediaTypeOrNull() }
            return guessed.getOrNull() ?: "image/jpeg".toMediaType()
        }

        override fun contentLength(): Long {
            return file.length()
        }

        override fun writeTo(sink: BufferedSink) {
            val length = file.length().toFloat()

            var lastTime = 0L
            FileInputStream(file).use { input ->
                // send first progress report.
                publishState(State.Uploading(progress = 0f))

                var sent = 0
                readStream(input) { buffer, len ->
                    sink.write(buffer, 0, len)
                    sent += len

                    val now = System.currentTimeMillis()
                    if (now - lastTime >= 100) {
                        lastTime = now

                        // send progress now.
                        val progress = sent / length
                        publishState(State.Uploading(progress = progress))
                    }
                }

                // tell that the file is sent
                publishState(State.Uploading(progress = 1f))
            }
        }
    }
}

fun isValidTag(tag: String): Boolean {
    val invalidTags = setOf("sfw", "nsfw", "nsfl", "nsfp", "gif", "video", "sound", "text")
    val invalid = tag.lowercase(Locale.ROOT) in invalidTags || tag.length < 2 || tag.length > 32
    return !invalid
}

fun isMoreRestrictiveContentTypeTag(tags: List<String>, tag: String): Boolean {
    val sorted = listOf("sfw", "nsfp", "nsfw", "nsfl")

    val newTagIndex = sorted.indexOf(tag.lowercase(Locale.ROOT))
    if (newTagIndex < 0) {
        return false
    }

    val maxExistingTagIndex = tags.map { sorted.indexOf(it.lowercase(Locale.ROOT)) }.maxOrNull()
    return maxExistingTagIndex == null || maxExistingTagIndex < newTagIndex
}

